larpandoracontent/LArTwoDReco/LArClusterAssociation/HitWidthClusterMergingAlgorithm.cc

```diff
@@ -43,12 +45,23 @@ void HitWidthClusterMergingAlgorithm::GetListOfCleanClusters(const ClusterList *
         if (LArHitWidthHelper::GetOriginalTotalClusterWeight(pCluster) < m_minClusterWeight)
             continue;
 
+        unsigned int numberOfProposedConstituentHits(LArHitWidthHelper::GetNProposedConstituentHits(pCluster, m_maxConstituentHitWidth, m_hitWidthScalingFactor));
+
+        // clusterSparseness [0 -> 1] where a higher value indicates sparseness
+        float clusterSparseness(1.f - (static_cast<float>(pCluster->GetNCaloHits()) / static_cast<float>(numberOfProposedConstituentHits)));
+
+        if (clusterSparseness < m_minClusterSparseness && pCluster->GetNCaloHits() != 1)
+            continue;
+
         m_clusterToParametersMap.insert(std::pair<const Cluster*, LArHitWidthHelper::ClusterParameters>(pCluster,
             LArHitWidthHelper::ClusterParameters(pCluster, m_maxConstituentHitWidth, false, m_hitWidthScalingFactor)));
 
         clusterVector.push_back(pCluster);
     }
 
+    if (clusterVector.empty())
+        return;
```

Comment:
So, the sort creates an instance of a class LArHitWidthHelper::SortByHigherXExtrema to sort the vector. The class has a member variable m_clusterToParametersMap which it uses to access the cluster parameters of the clusters. If this map is empty, a STATUS_CODE_NOT_INITIALIZED is thrown, so I return out of this function to stop this happening. I can remove this exception, since this would be caught later on in the sorting function anyway..

Reply:
Thanks for clarifying this subtlety! So this could be handled in several different ways. E.g.

1. If the sort function has nothing to do, regardless of the provision of the parameters map, it could return cleanly. This would protect all such usage.

2. The call to the sort function could have a try/catch around it (but be warned that raising exceptions is rather slow, and should only be done in exceptional circumstances).

3. The call to the sort function here could have the empty clusterVector check wrapping around it. And/or could just add a comment about why it's necessary to return.